Remove whitelist

AlexVlx opened the pull request:
This represents the Clang component of the work concerned with the removal
of the HCC whitelist. It changes the rule from "allow only some"
into "allow all except for" i.e. it starts from the assumption that all
language constructs are valid and prevents those that AMDGPU cannot
handle at the moment from generating LLVM IR. It does so by way of a
checker ADT, which gets called for all GlobalDecls. Some changes were
necessary in the exception handling / cleanup bits, to change to using
non address-space cast alloca-s for non programmer visible temporaries.
Similarly, when the compiler is executing the accelerator path the
guarded init for statics is not emitted - this is an important
divergence, but statics on accelerator have a troublesome state anyway,
and this does not affect them at all. The main benefit is that now, if
the code is inlineable and compatible with AMDGPU, it can be used
directly, without requiring adding [[hc]] or manually modifying the
whitelist.

AlexVlx committed:
This represents the Clang component of the work concerned with the
removal of the HCC whitelist. It changes the rule from "allow only some"
into "allow all except for" i.e. it starts from the assumption that all
language constructs are valid and prevents those that AMDGPU cannot
handle at the moment from generating LLVM IR. It does so by way of a
checker ADT, which gets called for all GlobalDecls. Some changes were
necessary in the exception handling / cleanup bits, to change to using
non address-space cast alloca-s for non programmer visible temporaries.
Similarly, when the compiler is executing the accelerator path the
guarded init for statics is not emitted - this is an important
divergence, but statics on accelerator have a troublesome state anyway,
and this does not affect them at all. The main benefit is that now, if
the code is inlineable and compatible with AMDGPU, it can be used
directly, without requiring adding [[hc]] or manually modifying the
whitelist.

AlexVlx committed:
This simplifies the usage of isWhiteListForHCC() and the rules for
emission when compiling in HC mode.

scchan commented on lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2009,77 +2011,93 @@ bool CodeGenModule::shouldOpportunisticallyEmitVTables() {
+      bool r = true;
+
+      if (x->isVariadic()) r = false;
+      if (x->isPure() || x->isVirtualAsWritten()) r = false;
```
any reason we would reject pure functions?

whchung replied:
@scchan this is to check if the function is a pure virtual function. I think it's fine here.

scchan replied:
yes, thanks for pointing out

whchung commented on lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -3294,6 +3302,10 @@ void CodeGenModule::EmitGlobalFunctionDefinition(GlobalDecl GD,
                                                    ForDefinition));
 
+  if (D->getAttr<CXXAMPRestrictAMPAttr>()) {
+    cast<llvm::Function>(GV)->addFnAttr("HC");
```
This is the main reason why in the PR for LLVM we only check for "HC" function attribute.

whchung commented on lib/CodeGen/CGDecl.cpp:
```diff
@@ -405,10 +405,10 @@ void CodeGenFunction::EmitStaticVarDecl(const VarDecl &D,
+  bool isHCCAcceleratorPath =
+      getLangOpts().CPlusPlusAMP && getLangOpts().DevicePath;
   // If this value has an initializer, emit it.
-  if (D.getInit() && !isCudaSharedVar && !isHCCTileStaticVar)
```
@AlexVlx would this change prevent from initializers of file-scope global variables be emitted?

AlexVlx replied:
This will prevent dynamic initializers from being emitted, this is correct. I do not believe that we actually support this correctly at the moment except for corner cases (and it is actually unclear what the meaning of it is in practice), and I do believe that we would benefit from correctly conveying that (as well as re-assessing the current mechanism for namespace scope variables in [[hc]]). Do we have any case which relies on dynamic initialisation of namespace scope variables referenced from [[hc]] and executes correctly at the moment (I could not identify them). I can re-enable it, but I am uncomfortable about its flakiness and it can create surprises / bugs, IMHO.

whchung commented on lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2089,26 +2107,16 @@ void CodeGenModule::EmitGlobalDefinition(GlobalDecl GD, llvm::GlobalValue *GV) {
-      if (!isWhiteListForHCC(*this, GD))
-        return;
-    } else {
-      // In host path:
```
@AlexVlx wouldn't these removed lines cause functions/variables with `restrict(amp)` be emitted in host path?

AlexVlx replied:
You are absolutely correct, that is an error on my part - thank you for pointing it out.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_whitelist

AlexVlx committed:
Address review comments by Jack, which outlined an error in the prior
commit that would have led to [[hc]] restricted code being emitted on
the host path.

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_whitelist

# Conflicts:
#	lib/CodeGen/CodeGenModule.cpp

AlexVlx committed:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_whitelist

# Conflicts:
#	lib/CodeGen/CodeGenModule.cpp

Commit:
Merge remote-tracking branch 'origin/feature_make_atomic_codegen_address_space_correct' into feature_remove_whitelist

Commit:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_whitelist

Commit:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_whitelist

AlexVlx committed:
This corrects code generation for expressions involving multiplication
and division of std::complex numbers on the accelerator path. The
existing disablement was somewhat unfortunate, and created issues as it
funneled floating point operands down the integer generation path, which
led to compiler crashes. The core issue stems from the use of dedicated
builtin library functions to handle these, namely __mul*c3 and __div*c3,
which live in the Compiler-RT project. We cannot use them since they are
neither inlineable nor fully accelerator viable, therefore at the moment
inline ersatz is provided. However, for robust std::complex support we
will have to consider extending Compiler-RT to include AMD-GCN-HSA
specific implementations of these functions, which correctly handle NaNs
or Infinities or any other special conditions, as specified by the C/C++
standards.

Commit:
Merge remote-tracking branch 'origin/clang_tot_upgrade' into feature_remove_whitelist
